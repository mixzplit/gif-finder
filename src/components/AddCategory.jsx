import { useState } from "react"

export const AddCategory = ({onNewCategory}) => {
  
  const [inputValue, setInputValue] = useState('');
  
  const onChangeHandler = (event) => {
    setInputValue(event.target.value);
  }

  const onSubmitHandler = (event) => {
    event.preventDefault(); // evitamos el refresh del form

    if(inputValue.trim().length <= 1) return;

    // Aqui obtenemos la categorias que se enviaron implicitamente
    // en la propiedades de componente el cual desestrucuramos en la 
    // linea 3 y actualizamos el estado con el valor del input y el
    // resto de categorias
    // setCategories(categories => [inputValue, ...categories]);


    onNewCategory(inputValue.trim());
    setInputValue(''); // limpiamos el input despues de enviar su valor
  }

  return (
    <form onSubmit={(e) => onSubmitHandler(e)}>
      <input type="text" placeholder="Buscar Gifs" value={inputValue} onChange={(e) => onChangeHandler(e)} />
    </form>
  )
}
