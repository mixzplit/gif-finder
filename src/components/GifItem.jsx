import  PropTypes from "prop-types";

// Desestructuramos el objeto que viene en las props
export const GifItem = ({ id, title, url }) => {
  return (
    <div className="card">
      <img src={url} alt={title} />
      <p>{title}</p>
    </div>
  );
};


GifItem.propTypes = {
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired
}