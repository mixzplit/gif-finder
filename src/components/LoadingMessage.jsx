import React from 'react'

export const LoadingMessage = ({isLoading}) => {
  console.log(isLoading);
  return (
    isLoading && ( <h2>Loading...</h2>)
  )
}
