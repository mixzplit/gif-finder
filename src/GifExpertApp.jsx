import React, { useState } from "react";
import { AddCategory, GifGrid } from "./components";

export const GifExpertApp = () => {
  const [categories, setCategories] = useState(['One Piece']);

  const onAddCategory = (newCategory) => {

    //Validamos que no se repitan las categorias
    if(categories.includes(newCategory)) return;

    //Destructuramos y usamos el operador spread para
    //obtener el estado actual de las categorias y
    // agregamos la nueva
    setCategories([newCategory, ...categories ]);
  }



  return (
    <>
      <h1>GifExpertApp</h1>

      {/* Input */}
      {/* <AddCategory setCategories={ setCategories } /> */}
      <AddCategory onNewCategory={ (e) => onAddCategory(e) } />

      {/* Listado de GIF */}
        {
          categories.map((category) => (
            <GifGrid key={category} category={category} />
          ))
        }
    </>
  );
};
