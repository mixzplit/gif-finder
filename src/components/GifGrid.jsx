import { useFetchGifs } from "../hooks/useFetchGifs";
import { GifItem } from "./GifItem";
import { LoadingMessage } from "./LoadingMessage";
import  PropTypes from "prop-types";


export const GifGrid = ({ category }) => {
  const { images, isLoading } = useFetchGifs(category);

  //console.log({ isLoading });

  return (
    <>
      <h3>{category}</h3>
      <LoadingMessage isLoading={isLoading} />
      <div className="card-grid">
        {images.map((img) => {
          // con el operador spread esparcimos todo el objeto
          // al componente
          return <GifItem key={img.id} {...img} />;
        })}
      </div>
    </>
  );
};

GifGrid.propTypes = {
  category: PropTypes.string.isRequired
}