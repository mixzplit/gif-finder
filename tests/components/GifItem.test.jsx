import { render, screen } from "@testing-library/react";
import { GifItem } from "../../src/components";

describe('Unit test for <GifItem />', () => {

  const title = 'One Piece';
  const url = 'https://prueba.sf.ar/';

  test("Debe hacer match con el snapshot", () => {
    const { container } = render(<GifItem title={title} url={url} />)
    expect(container).toMatchSnapshot(); 
  });

  test('Debe de mostrar la URL y texto ALT', () => {
    render(<GifItem title={title} url={url} />);
    screen.debug();
    // Desestructuramos
    const {src, alt} = screen.getByRole('img');
    expect(src).toBe(url);
    expect(alt).toBe(title);

  });

  test('Debe de mostrar el titulo del componente', () => {
    render(<GifItem title={title} url={url} />);
    expect(screen.getByText(title)).toBeTruthy();
  });
});