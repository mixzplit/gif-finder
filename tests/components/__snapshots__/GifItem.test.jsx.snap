// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit test for <GifItem /> Debe hacer match con el snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="One Piece"
      src="https://prueba.sf.ar/"
    />
    <p>
      One Piece
    </p>
  </div>
</div>
`;
